docs/proposals/device-crd.md

```diff
@@ -0,0 +1,554 @@
+  // A list of visitors describing the properties which can be accessed with the protocol.
+  // Required.
+  Visitors     []PropertyVisitor `json:"visitors,omitempty"`
+}
+
+// ProtocolType describes the protocol type used to communicate with the device instance.
+type ProtocolType string
+
+const (
+	// Bluetooth Low Energy(BLE) Protocol
+	BluetoothLE  ProtocolType = "BluetoothLE"
+	// Zigbee Protocol
+	Zigbee       ProtocolType = "Zigbee"
+	// Modbus Protocol
+	Modbus       ProtocolType = "Modbus"
+	// BACnet Protocol
+	BACnet       ProtocolType = "BACnet"
+	// OPC UA Protocol
+	OPCUA        ProtocolType = "OPCUA"
+)
+
+// PropertyVisitor contains details like which property can be accessed and the
+// access mechanisms are described via key-value pairs in the metadata.
+type PropertyVisitor struct {
+  // The device property visitor name.
+  // Required.
+  Name         string            `json:"name,omitempty"`
+  // The device property name to be accessed. This should refer to one of the
+  // device properties defined in the device model.
+  // Required.
+  PropertyName string            `json:"propertyName,omitempty"`
+  // Additional metadata about the how to access the device property.
+  // Required.
+  Metadata     map[string]string `json:"metadata,omitempty"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DeviceModel is the Schema for the device model API
+// +k8s:openapi-gen=true
+type DeviceModel struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec DeviceModelSpec `json:"spec,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DeviceModelList contains a list of DeviceModel
+type DeviceModelList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []DeviceModel `json:"items"`
+}
+```
+
+### Device model sample
+```yaml
+apiVersion: devices.kubeedge.io/v1alpha1
+kind: DeviceModel
+metadata:
+  labels:
+    description: 'TI Simplelink SensorTag Device Attributes Model'
+    manufacturer: 'Texas Instruments'
+    model: CC2650
+  name: sensor-tag-model
+spec:
+  properties:
+  - name: temperature
+    description: 'Temperature from Temperature Sensor in Degrees Fahrenheit'
+    propertyConfig:
+      dataType: float
+      accessMode: r
+      unit: 'degree fahrenheit'
+  - name: temperature-enable
+    description: 'Enable Temperature Sensor'
+    propertyConfig:
+      dataType: string
+      accessMode: rw
+  propertyVisitors:
+    - protocol:  BluetoothLE # valid values : [BluetoothLE, Zigbee, ModbusTCP, ModbusRTU, BACnet, OPCUA]
+      visitors:
+      - name: temperature
+        propertyName: temperature
+        visitorConfig:
+          uuid: F000AA01-0451-4000-B000-000000000000
+      - name: temperatureEnable
+        propertyName: temperature-enable
+        visitorConfig:
```

Comment:
visitorConfig -> visitConfigs? means we are going to visit the configs?

Reply:
in this example , it means that if you want to enable data collection for temperature from a sensor tag, you need to access a register with a specific uuid. there could be many more details other than uuid like at what index , offset to start reading data , what is the byte-size of the data etc.

Reply:
As there must be multiple configurations, I suggest add `s` in this field. e.g `visitorConfigs` or `visitConfigs`

Reply:
updated APIs don't have this struct